File: packages/domain/src/preset.ts
/** Preset kind — locomotion or secondary motion */
export type MotionPresetKind = 'locomotion' | 'secondary_motion';

/** An anchor definition within a preset */
export interface PresetAnchor {
  name: string;
  kind: string;
  parentName: string | null;
  falloffWeight: number;
  hintX: number;
  hintY: number;
}

/** Template/intent settings captured in a preset */
export interface PresetMotionSettings {
  intent?: string | null;
  templateId?: string | null;
  direction?: string | null;
  strength?: number | null;
  frameCount?: number | null;
  phaseOffset?: number | null;
}

/** Full preset document */
export interface MotionPresetDocument {
  schemaVersion: number;
  id: string;
  name: string;
  kind: MotionPresetKind;
  description?: string | null;
  anchors: PresetAnchor[];
  motionSettings: PresetMotionSettings;
  targetNotes?: string | null;
  createdAt: string;
  modifiedAt: string;
}

/** Lightweight preset summary for listing */
export interface MotionPresetSummary {
  id: string;
  name: string;
  kind: MotionPresetKind;
  description?: string | null;
  anchorCount: number;
  hasHierarchy: boolean;
  templateId?: string | null;
  createdAt: string;
  modifiedAt: string;
}

/** Result from saving a preset */
export interface PresetSaveResult {
  id: string;
  name: string;
}

/** Compatibility check result */
export interface PresetCompatibility {
  tier: 'compatible' | 'partial' | 'incompatible';
  matchingAnchors: string[];
  missingAnchors: string[];
  extraAnchors: string[];
  wouldExceedLimit: boolean;
  notes: string[];
}

/** Optional overrides for preset apply — does not modify the saved preset */
export interface PresetApplyOverrides {
  strength?: number;
  direction?: string;
  frameCount?: number;
  phaseOffset?: number;
}

/** Result from applying a preset */
export interface PresetApplyResult {
  createdAnchors: string[];
  updatedAnchors: string[];
  skipped: string[];
  warnings: string[];
  appliedSettings?: PresetMotionSettings | null;
}

/** Per-frame result within a batch apply */
export interface BatchFrameResult {
  frameIndex: number;
  frameId: string;
  created: number;
  updated: number;
  skipped: number;
  warnings: string[];
}

/** Result from batch-applying a preset across frames */
export interface BatchApplyResult {
  totalFrames: number;
  appliedFrames: number;
  skippedFrames: number;
  perFrame: BatchFrameResult[];
  summary: string[];
  appliedSettings?: PresetMotionSettings | null;
}

/** Per-anchor diff in a preview */
export interface PresetAnchorDiff {
  name: string;
  action: 'create' | 'update' | 'skip';
  changes: string[];
}

/** Non-mutating preview of what applying a preset would do */
export interface PresetPreviewResult {
  presetName: string;
  presetKind: MotionPresetKind;
  anchorDiffs: PresetAnchorDiff[];
  effectiveSettings: PresetMotionSettings;
  warnings: string[];
  scopeFrames: number;
}

File: apps/desktop/src-tauri/src/engine/preset.rs
use serde::{Deserialize, Serialize};

// ─── Preset domain types ─────────────────────────────────────

/// Preset kind — locomotion or secondary motion.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MotionPresetKind {
    Locomotion,
    SecondaryMotion,
}

/// An anchor definition within a preset.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PresetAnchor {
    pub name: String,
    pub kind: String,
    pub parent_name: Option<String>,
    pub falloff_weight: f32,
    /// Suggested relative position (0.0–1.0 normalized to canvas).
    #[serde(default)]
    pub hint_x: f32,
    #[serde(default)]
    pub hint_y: f32,
}

/// Template/intent settings captured in a preset.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PresetMotionSettings {
    /// For locomotion: intent ID (e.g. "idle_bob", "walk_cycle_stub").
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub intent: Option<String>,
    /// For secondary: template ID (e.g. "wind_soft", "hanging_swing").
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub template_id: Option<String>,
    /// Direction preference.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub direction: Option<String>,
    /// Strength (secondary only, 0.1–2.0).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub strength: Option<f64>,
    /// Frame count.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub frame_count: Option<u32>,
    /// Phase offset (secondary only, 0–TAU).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub phase_offset: Option<f64>,
}

/// Full preset document — serialized to disk.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MotionPresetDocument {
    /// Schema version for forward compatibility.
    pub schema_version: u32,
    pub id: String,
    pub name: String,
    pub kind: MotionPresetKind,
    /// Optional short description.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Anchor structure (names, kinds, hierarchy, falloff).
    pub anchors: Vec<PresetAnchor>,
    /// Motion template/intent settings.
    pub motion_settings: PresetMotionSettings,
    /// Human-readable notes about recommended targets.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub target_notes: Option<String>,
    /// Created timestamp (ISO 8601).
    pub created_at: String,
    /// Last modified timestamp (ISO 8601).
    pub modified_at: String,
}

impl MotionPresetDocument {
    pub const CURRENT_SCHEMA_VERSION: u32 = 1;

    pub fn new(name: String, kind: MotionPresetKind) -> Self {
        let now = chrono_now();
        Self {
            schema_version: Self::CURRENT_SCHEMA_VERSION,
            id: uuid::Uuid::new_v4().to_string(),
            name,
            kind,
            description: None,
            anchors: Vec::new(),
            motion_settings: PresetMotionSettings {
                intent: None,
                template_id: None,
                direction: None,
                strength: None,
                frame_count: None,
                phase_offset: None,
            },
            target_notes: None,
            created_at: now.clone(),
            modified_at: now,
        }
    }
}

/// Summary for listing presets (lightweight, no full anchor data).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MotionPresetSummary {
    pub id: String,
    pub name: String,
    pub kind: MotionPresetKind,
    pub description: Option<String>,
    pub anchor_count: u32,
    pub has_hierarchy: bool,
    pub template_id: Option<String>,
    pub created_at: String,
    pub modified_at: String,
}

impl MotionPresetDocument {
    pub fn to_summary(&self) -> MotionPresetSummary {
        MotionPresetSummary {
            id: self.id.clone(),
            name: self.name.clone(),
            kind: self.kind,
            description: self.description.clone(),
            anchor_count: self.anchors.len() as u32,
            has_hierarchy: self.anchors.iter().any(|a| a.parent_name.is_some()),
            template_id: self.motion_settings.template_id.clone()
                .or_else(|| self.motion_settings.intent.clone()),
            created_at: self.created_at.clone(),
            modified_at: self.modified_at.clone(),
        }
    }
}

/// Simple timestamp (no chrono dependency — use basic format).
fn chrono_now() -> String {
    // Use system time for a basic ISO-ish timestamp
    let now = std::time::SystemTime::now();
    let dur = now.duration_since(std::time::UNIX_EPOCH).unwrap_or_default();
    let secs = dur.as_secs();
    // Convert to basic UTC-like format: seconds since epoch
    // For a real ISO 8601 we'd need chrono, but this is sufficient for ordering
    format!("{}", secs)
}
